Made ByteSynkArray to use shared_ptr

Trial97 opened the pull request:
<!--
Hey there! Thanks for your contribution.

Please make sure that your commits are signed off first.
If you don't know how that works, check out our contribution guidelines: https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md#signing-your-work
If you already created your commits, you can run `git rebase --signoff develop` to retroactively sign-off all your commits and `git push --force` to override what you have pushed already.

Note that signing and signing-off are two different things!
-->
ok this should fix https://discord.com/channels/1031648380885147709/1032130809055940690/threads/1118963246330810469 and maybe https://github.com/PrismLauncher/PrismLauncher/issues/1199

Trial97 committed:
Made ByteSynkArray to use shared_ptr

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Ryex commented:
when I attempted this change I ended up with leaked shared pointers because of cyclic references which is why it wasn't part of the initial memory leak fixes. Have you checked with address sanitizer that these don't leak?

Trial97 commented:
Yeah,leaks:
```
SUMMARY: AddressSanitizer: 1506594 byte(s) leaked in 6322 allocation(s).
```
Let me tweak more.

Trial97 commented:
@Ryex can you help me with the address sanitizer?
I added the flags for the build: `set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address -fno-omit-frame-pointer")` and added this env var: `export ASAN_OPTIONS=fast_unwind_on_malloc=0:detect_stack_use_after_return=1:check_initialization_order=1:strict_init_order=1:malloc_context_size=100 `
Now the question is how do I determine if I have a problem with the shared_ptr?
My initial test case was:
1. open the generated executable
2. add a new instance
3. select all the mods8 (first mod pack from curse forge)
4. go to the process of adding the mod pack
5. after all is done just close the prism(completely)
6. on close I was greeted with some leaks:`Indirect leak of 34 byte(s) in 3 object(s) allocated from:`

My problem is if I apply the same to develop I get similar results.
Should I open an issue with this?
*Also this is not limited to curseforge mod packs: except modrinth, all providers have similar results for the analyzer(same steps)

Ryex commented:
if your leaking the shared pointer you should get stack traces leading under `make_shared` when you exit

shared pointers leak when there is memory behind a shared pointer which holds a shared pointer that either directly or indirectly holds a shared pointer to the parent. In theory the solution is using weak pointers at some point in that chain to break the cycle.

I really did try my best to remove all the leaks when I made that PR but it could be possible new ones were introduced or that I missed some.

---

The trick to working with the address sanitizer, especially with indirect leaks is to see where the memory that leaks was initialized and try to trace the pointer mack to it's nearest owner. then go through and try to find where it should be drooped. if that pointer traces back to the ownership of a shared pointer try to find the cycle.

the solution may be that we need to start using weak pointers for instances passes to dialogs or pages out of their ownership in the Application class.

---

The other place to be extra careful with shared pointers is anywhere they may be captured by a lambda that is connected to a signal or slot. if you can, only capture the raw pointer using `get` or explicitly capture by reference so that a copy isn't made increasing the ref-count.

Trial97 commented:
Ok so I give up regarding the address sanitizer thing.
I tested: all resource download, no leak there; all modpacks download, except modrinth all the other providers has some leak; also tested the mod update, and found leaks there too.
But no leaks related to `std::make_shared<QByteArray>` or `std::shared_ptr<QByteArray>` and all the leaks I was able to reproduce them on develop.
***
Also for me, the following character creates a leak: `✔`(https://github.com/PrismLauncher/PrismLauncher/blob/fd9a8d1551e5736aff6da10d8d00cc631c6d6ee0/launcher/ui/dialogs/BlockedModsDialog.cpp#L166) removing it the leak is fixed :exploding_head:

Trial97 committed:
Fixed leaks

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into net_job_crash

Trial97 committed:
Removed extra headers

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Ok status update: well fixed the found leaks in modpacks download(If needed I can create a separate PR with only them)
Except the one caused by the character mention previously.
Did not find any leak to lead to the ByteSyncArray, or associated job. (tested on all resource downloads)

Ryex reviewed (approved):
LGTM

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into net_job_crash

TheKodeToad commented on launcher/modplatform/helpers/HashUtils.h:
```diff
@@ -21,6 +22,9 @@ class Hasher : public Task {
     QString getPath() const { return m_path; };
 
+   signals:
+    void getResults(QString hash);
```
Maybe `resultsReady` would be a better name? This looks like a getter

TheKodeToad commented on launcher/modplatform/atlauncher/ATLPackInstallTask.cpp:
```diff
@@ -99,11 +99,12 @@ void PackInstallTask::onDownloadSucceeded()
-        qWarning() << "Error while parsing JSON response from ATLauncher at " << parse_error.offset << " reason: " << parse_error.errorString();
-        qWarning() << response;
+    QJsonParseError parse_error{};
+    QJsonDocument doc = QJsonDocument::fromJson(*response.get(), &parse_error);
```
Afaik, `.get` shouldn't be necessary here

TheKodeToad reviewed (approved):
\<tired review\> overall good

Trial97 committed:
Renamed getResults to resultsReady

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Merge branch 'develop' of https://github.com/PrismLauncher/PrismLauncher into net_job_crash
